app/src/org/commcare/android/session/DevSessionRestorer.java

```diff
@@ -47,11 +62,75 @@ public static void tryAutoLoginPasswordSave(String password) {
         }
     }
 
-    private static boolean autoLoginEnabled() {
+    public static boolean autoLoginEnabled() {
         return BuildConfig.DEBUG && DeveloperPreferences.isAutoLoginEnabled();
     }
 
     public static void clearPassword(SharedPreferences prefs) {
         prefs.edit().remove(CommCarePreferences.LAST_PASSWORD).commit();
     }
+
+    /**
+     * Builds a session object using serialized data stored in the app
+     * preferences. Currently doesn't support restoring the session stack, only
+     * the current frame.
+     */
+    public static AndroidSessionWrapper restoreSessionFromPrefs(CommCarePlatform platform) {
+        SharedPreferences prefs =
+                CommCareApplication._().getCurrentApp().getAppPreferences();
+        String serializedSession = prefs.getString(CommCarePreferences.CURRENT_SESSION, null);
+        if (serializedSession != null) {
+            try {
+                byte[] sessionBytes = Base64.decode(serializedSession, Base64.DEFAULT);
+                DataInputStream stream =
+                        new DataInputStream(new ByteArrayInputStream(sessionBytes));
+
+                CommCareSession restoredSession =
+                        CommCareSession.restoreSessionFromStream(platform, stream);
+
+                Log.i(TAG, "Restoring session from storage");
+                return new AndroidSessionWrapper(restoredSession);
+            } catch (Exception e) {
+                clearSession(prefs);
+                Log.w(TAG, "Restoring session from serialized file failed");
+            }
+        }
+
+        // no saved session or failed to restore; return a blank session
+        return new AndroidSessionWrapper(platform);
+    }
+
+    /**
+     * Save the current session frame to app shared preferences.
+     */
+    public static void saveSessionToPrefs() {
+        CommCareApp ccApp = CommCareApplication._().getCurrentApp();
+        if (ccApp == null) {
+            return;
+        }
+
+        SharedPreferences prefs = ccApp.getAppPreferences();
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        DataOutputStream serializedStream = new DataOutputStream(baos);
+        try {
+            CommCareApplication._().getCurrentSession().serializeSessionState(serializedStream);
+        } catch (IOException e) {
+            Log.w(TAG, "Failed to serialize session");
+            return;
+        } catch (SessionStateUninitException e) {
+            Log.w(TAG, "Attempting to save a non-existent session");
+            return;
+        }
+        String serializedSession = Base64.encodeToString(baos.toByteArray(), Base64.DEFAULT);
+        prefs.edit().putString(CommCarePreferences.CURRENT_SESSION, serializedSession).commit();
```

Comment:
Does it matter that we are storing unencrypted session info in the app preferences?